### PR TITLE
Make injection more generic

### PR DESCRIPTION
**Pull request by FasterSpeeding**
The aim of this PR is to remove any dependency on tanjun.abc.Context from injection logic and introduce an injection Context to allow for more flexibility (letting me also implement features like injection in places where tanjun.abc.Context isn't present and caching the results of callbacks)

Update: adding injection support to event dispatch is gonna be saved for another PR

### CODE REVIEW

File: tanjun/__init__.py [hunk 1]
```diff
@@ -85,9 +85,8 @@ async def hello(ctx: tanjun.abc.Context, user: typing.Optional[hikari.User]) ->
 # will have to be started after the Hikari client.
 async def main() -> None:
     await bot.start()
-    await client.open()
-    await bot.join()
-    await client.close()
+    async with client.open()
```
Comment (nxtlo) [inline tanjun/__init__.py]:
```suggestion
    async with client.open():
```

**Comment by FasterSpeeding** [comment]
~~Note to self, rename special cased type injector and type injector callbacks~~ actually no just hide it as internal detail in most places but still include `def _` method in  pyi

**Comment by FasterSpeeding** [comment]
Things to do
- [x] Do we want to pass around args and kwargs in injecting stuff as 2 arguments instead of *args and **kwargs (this is unnecessary)
- [x] BaseInjectableValue InjectableValue and  rename to BaseInjectableCallback InjectableCallback
- [x] Add the command context types to the context's special cases if that isn't already being done

File: tests/test_injecting.py [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+
+class TestUndefined:
+    def test_bool_operator(self):
+        assert bool(tanjun.injecting.UNDEFINED) is False
+
+    def test_undefined_is_singleton(self):
+        assert tanjun.injecting.Undefined() is tanjun.injecting.UNDEFINED
+        assert (
+            tanjun.injecting.Undefined() is type(tanjun.injecting.UNDEFINED)()  # noqa: E721
+        )  # noqa: E721 # false-positive about comparing types
```
Comment (FasterSpeeding) [inline tests/test_injecting.py]: Yeet this comment

**Pull request merged**